[cire] step-6 머스타치 템플릿 엔진 모방하여 구현 및 세션, 인터셉터 구현

cire0304 opened the pull request:
# 느낀점

하나의 기능을 구현을하고, 하나의 커밋을 해야한다는 사실을 뼈저리게 느꼈습니다.

하나의 기능을 구현하다가, 또 다른 것도 구현하고 싶어서 하다가, 또 다른 기능을 동시에 구현하다가 커밋하는 타이밍을 놓쳐버렸습니다.
그러다가 버그가 발생했고, 깃의 히스토리 기능도 쓰지 못해서 정신이 나가는 줄 알았습니다.

결과적으로는 한줄한줄 디버깅을 했고, 버그는 찾아서 해결했습니다.
앞으로는 정말 커밋을 잘해야 한다는 것을 느낄 수 있는 미션이었습니다...

# 아쉬운 점 1

머스타치 템플릿 엔진의 기능을 모방하여 서버 사이드 렌더링 기능을 구현했습니다.

하지만, 현재 코드는 거의 암호화된 수준입니다.

시간 남을때, 리펙토링 하려 합니다. (하지만, 이제 수정할 일 없을 것 같은데 리펙토링을 해야할까요?)

# 아쉬운 점 2

테스트 케이스를 만들면서 하지 않았습니다.

예전에 테스트 케이스를 만들어 둔 기능은 버그가 발생해도, 빠르게 디버깅이 가능했습니다.
하지만 그렇지 않은 기능은 정말, 한줄 한줄 디버깅을 하여 수정해야합니다.

앞으로는 정말 기능 하나당 커밋하나 테스트 하나, 이런 식으로 개발해야 겠습니다.
(아니면 이번 주말에, 현재 까지 구현한 기능의 테스트 케이스를 만드는 것도 좋은 연습이 될 것 같습니다.)

# View 클래스를 아쉽게 구현했다.

스프링의 핸들러 반환 값이 문자열 타입의 ViewName 일 때, 자동으로 파일을 찾아가 렌더링을 합니다.

하지만 현제 저의 코드는, 확장자 명까지 포함된 파일 이름을 viewName으로 반환해야 합니다.

나중에 리펙토링 하려 합니다.

# 아쉽지만 안할 것 같은 것

현재 머스타치 문법을 적용한 HTML 은 index.html, user/list.html 두 개 뿐 입니다.

다른 HTML 도 동적으로 만들면 좋을 것 같지만... 나중에 하려 합니다!

cire0304 committed:
Rename : 이름 변경

클래스는 명사, 메소드는 동사로 이름을 짓는게 좋다.

cire0304 committed:
Style : 가독성 향상을 위한 코드 스타일 변경

cire0304 committed:
Feat : 편의 메소드 작성

클라이언트에서 보내온 파라미터를 리턴하는 편의 메소드 작성

cire0304 committed:
Feat : 뷰 및 뷰 리졸버 기능 구현

클라이언트에게 응답할 때, 2가지 경우가 있다.
1. 리다이렉트
2. 그 외
따라서 2가지 경우에 대해 각각 뷰를 만들어주었다.

뷰 리졸버는 요청에 맞는 뷰를 찾는 기능으로 구현하였다.

cire0304 committed:
Refactor : 뷰 및 뷰 리졸버 기능을 적용하도록 코드 수정

cire0304 committed:
Chore : 패키지 이동

cire0304 committed:
Feat : 서버사이드 탬플릿 엔진 구현

머스타치 문법을 가지는 템플릿 엔진을 구현하였다.

하지만 리펙토링이 필요하다.

cire0304 committed:
Test : 템플릿 엔진 테스트 코드 작성

cire0304 committed:
Chore : 라이브러리 변경

cire0304 committed:
Update : MylittleMustache 반전 섹션 기능 업데이트

하지만 코드 정리를 해야한다..

cire0304 committed:
Refactor : model 클래스 리펙토링

Model 클래스 대신 Map<>을 사용하여, ModelAndView 클래스로 변경

cire0304 committed:
Feat : 섹션 기능 활용을 위한 인터셉터 기능 구현

cire0304 committed:
Feat : 스프링의 뷰 리졸버 기능 모방하여 구현

cire0304 committed:
Remove : 사용되지 않는 클래스 제거

cire0304 committed:
Refactor : HTTP 응답 헤더를 위한 ContentType Enum 추가

cire0304 committed:
Feat : 스프링의 ReqeustMappingHandlerMapping 및 ArgumentReolver 기능 모방하여 구현

스프링의 HandlerMapping의 동작 방식을 보고, ArgumentReolver 같은 기능이 있으면 편할 것 같아 구현했다.

cire0304 committed:
Fix : HttpRequest 쿠키 버그 픽스

요청에서 쿠키값이 없을 때, HttpRequest 쿠키 필드가 널이어서, 널포인트 예외 발생할 가능성이 있는 문제 픽스

cire0304 committed:
Refactor : MethodReolver 기능 적용

cire0304 committed:
Refactor : 클린코드 지향하여 리펙토링

cire0304 committed:
Config : 객체간의 관계 재정의

cire0304 committed:
Remove : 사용되지 않는 파일 제거

cire0304 committed:
Chore : 사용되지 않는 라이브러리 제거

cire0304 committed:
Feat : 로그아웃 기능 구현

cire0304 committed:
Feat : 사용자 목록 페이지 출력 구현

cire0304 committed:
FIx : 리다이렉트 버그 픽스

리다이렉트 필드 변수 초기화를 안하고 있었다.

cire0304 committed:
Feat : 권한이 없는 페이지 요청 시 거절 기능 구현

cire0304 committed:
Chore : EOL 추가

honux77 commented on src/main/java/configure/AppConfiguration.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.HashMap;
+
+// TODO : 현재는 싱글톤 처럼 동작하지는 않는다.
+public class AppConfiguration {
```
알고 있겠지만 사실 모든 빈이 항상 싱글톤으로 동작하는 건 아닙니다.  https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-factory-scopes 참고해 보세요.

honux77 commented on src/main/java/interceptor/DefaultInterceptor.java:
```diff
@@ -0,0 +1,32 @@
+            return true;
+        }
+
+        modelAndView.addModel("user", user);
+        modelAndView.addModel("login", true);
+        return true;
```
기본 인터셉터의 실행범위는 잘 모르겠지만 모델엔뷰를 여기서 추가하는 게 살짝 어색하게 느껴집니다.

honux77 reviewed (commented):
안녕하세요? 시레 늦어서 죄송합니다 ㅜㅜ. 컨디션이 저조해서...
코멘트처럼 커밋을 자주 하는 습관은 엄청 중요하니 다음에는 지키시고, 그리고 브랜치도 잘 활용하면 좋습니다. 지금처럼 branchV6Copy 와 같은 이상한 이름은 지양해 주세요. 

계속 다양한 고민을 하고 시도해서 참 좋은 것 같네요. 
필수는 아니지만 7단계 정도는 마무리 지으면 좋을 것 같아요.

honux77 commented on build.gradle:
```diff
@@ -5,7 +5,7 @@ plugins {
 group 'codesquad'
 version '1.0-SNAPSHOT'
-sourceCompatibility = 11
+sourceCompatibility = 15
```
엇? JDK를 15로 바꿨는데 이유가 있을까요? 특별한 이유가 없다면 11을 사용해 주세요.

honux77 commented on src/main/java/util/MyHandlerMethod.java:
```diff
@@ -0,0 +1,43 @@
+import java.lang.reflect.Method;
+import java.util.Collection;
+
+public class MyHandlerMethod {
```
클래스 이름이 너무 어색해요. 핸들러와 메소드가 있어서 이렇게 이름을 지은 거 같은데 더 좋은 이름이 없을까요?  ControllerAndMethod 어때요? ㅋㅋ 
그리고 결국은 URL -> method로의 최종 매핑이 필요해서 클래스와 method를 가져오는데, 람다를 사용하면 더 깔끔하게 처리할 수 있지 않을까 생각이 듭니다.

honux77 reviewed (commented):
참 그리고 RequestMapping에서     
`public static Method requestHandlerMapping(Class<? extends Controller> controller, ...`
가 조금 더 명확하지 않을까요?

honux77 commented on src/main/java/servlet/DispatcherServlet.java:
```diff
@@ -1,59 +1,61 @@
-        String pathInfo = httpRequest.getPathInfo();
-        String requestMethod = httpRequest.getMethod();
+        // 요청에 맞는 컨트롤러와 메소드를 찾는다.
+        MyHandlerMethod handlerMethod = new MyHandlerMethod(httpRequest, controllers);
```
수업 때 말한 것처럼 콘트롤러가 default 인 경우 modelAndView를 만들지 않고 바로 처리하도록 하는 게 좋을 것 같아요.

honux77 commented on src/main/java/interceptor/Interceptor.java:
```diff
@@ -0,0 +1,9 @@
+import view.ModelAndView;
+
+public interface Interceptor {
+    public boolean preHandle(HttpRequest httpRequest, HttpResponse httpResponse, ModelAndView modelAndView);
```
인터셉터도 마찬가지로 static URL 요청에는 실행하지 않도록 하는 로직의 추가가 필요해 보입니다.

honux77 reviewed (commented):
참 전에도 언급한 거 같은데 세션매니저가 Cookie 를 가지는 건 맞지 않아요.
Cookie는 단순히 HttpRequest에 있는 특정 필드일 뿐이고
서버에서는 세션을 유지해야 합니다. 이 부분은 꼭 수정해 주세요. (언제?)

honux77 reviewed (changes requested):
시레~ 요구사항 구현하느라 수고 많으셨어요. 늦은 리뷰 죄송하고 몇가지 그래도 중요한 피드백을 남겼으니 반영해 주면 좋을 것 같습니다.

cire0304 commented:
> 시레~ 요구사항 구현하느라 수고 많으셨어요. 늦은 리뷰 죄송하고 몇가지 그래도 중요한 피드백을 남겼으니 반영해 주면 좋을 것 같습니다.

넵 알겠습니다!

cire0304 committed:
Refactor : 인터셉터에서 메서드어뎁터로 모델엔뷰 이동

마스터의 피드백을 반영하여, 모델엔뷰의 위치를 이동하였다.
모델엔뷰의 모델 추가는 핸들러의 공통적인 로직으로 생각하였다.
따라서, 핸들러를 실행하는 책임을 가지고 있는 어뎁터에서 모델 데이터를 추가하도록 하였다.

cire0304 committed:
Rename : MyHandlerMethod 이름 변경

스프링 MVC에서 Handler는 일반적으로 컨트롤러와 그 안에 있는 메서드를 지칭한다고 한다.
이전의 이름 MyHandlerMethod 에서 Method 부분이 의미가 중복되므로 제거하였다.

cire0304 committed:
Feat : static 파일을 요청할 경우, 로직없이 응답하도록 코드 수정

기존의 디폴트 컨트롤러의 역활을 static 파일 요청에 응답하는 것으로 변경하였다.

cire0304 committed:
Feat : template 요청일 경우 UserController가 응답하도록 코드 수정

cire0304 commented:
> 참 전에도 언급한 거 같은데 세션매니저가 Cookie 를 가지는 건 맞지 않아요. Cookie는 단순히 HttpRequest에 있는 특정 필드일 뿐이고 서버에서는 세션을 유지해야 합니다. 이 부분은 꼭 수정해 주세요. (언제?)

호눅스! 질문이 있습니다.
저는 현재 코드에서 세션매니저가 1) 클라이언트 객체에서 쿠키 객체를 가져오고, 2) 쿠키 객체에서 쿠키 값을 읽어 섹션을 가져오는 것으로 이했습니다.

제가 궁금한 점은 섹션매니저의 필드에 쿠키 객체는 없지만, 메소드 실행 중 쿠키 객체를 사용하게 되는 것도, 세션매니저가 Cookie를 가지고 있게 되는 것인지 궁금합니다!

honux77 commented:
시레~ 섹션 (X) 세션 (O) 글 쓰고 나면 한 번 더 맞게 썼는지 검토하는 습관을 가지면 좋을 거 같아요. 
코드를 다시 읽어보니 제가 잘못 이해했네요. 세션 코드에는 아무 문제가 없네요 ㅜㅜ. 죄송합니다.

cire0304 commented:
> 시레~ 섹션 (X) 세션 (O) 글 쓰고 나면 한 번 더 맞게 썼는지 검토하는 습관을 가지면 좋을 거 같아요. 코드를 다시 읽어보니 제가 잘못 이해했네요. 세션 코드에는 아무 문제가 없네요 ㅜㅜ. 죄송합니다.

아닙니다 호눅스!
오히려 항상 제 읽기 어려운 코드(분발하겠습니다 ㅜㅜ)를  리뷰해주셔서 감사합니다.

그리고 글을 쓰고 검토하는 습관을 가지도록 하겠습니다!!
